metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1909,6 +1921,31 @@ class MetalsLanguageServer(
         ammonite.start().asJavaObject
       case ServerCommands.StopAmmoniteBuildServer() =>
         ammonite.stop()
+
+      case ServerCommands.StartScalaCliServer() =>
```

Comment:
Quick question before you get super far on this, but why is this needed? Seeing that `scala-cli` has a `.bsp/` and is functioning as a build server, wouldn't it be preferable that we treat it as such instead of special casing it? For example we already have a `ServerCommands.ConnectBuildServer()` and a `ServerCommands.DisconnectBuildServer()`. Couldn't those be re-used for this? If possible I think we should avoid even more commands if others can be re-used.

Reply:
Seeing that there is some more action on this pr now I wanted to follow-up on this. I still really wish this wasn't needed. I know we do this for Ammonite, but I honestly wish these didn't exist. I've two main points for this.

1. These command are really difficult to discover. I'd argue that most users don't know that the "start ammonite server" command even exists. Instead, they just rely on the pop up from metals that it recognizes a script and asks the user to import it.
2. This also adds in more mental overhead for the user to keep in mind. Do I need to stop the build server? Do I need to manually stop scala-cli?

As an alternative in the situation you mentioned would it be possible that if a user has an sbt/mill/whatever build with a script that when they open the script it prompts the user:

```
Script detected. What type of script is this?
1. Ammonite
2. scala-cli
```

Then if the user selects 2 under the hood we run `setup-ide` which will create the `.bsp` entry and then we can just switch to it to offer support. Then we can just fully rely on BSP discovery and the existing commands that we have to stop the build server, restart the build server, or switch the build server. This would use existing logic, and add no more mental overhead of new commands for the user. Is something like this possible?

Reply:
The commits I just pushed should partially address some of your points. I'll discuss that in more detail here tomorrow.

Reply:
The code pushed since your comment now opens a popup similar to the one you're describing. Upon opening an `.sc` file, users are asked whether they want to load it as a Scala CLI or an Ammonite script (if they accept either, a second popup asks if they automatically want to do so in the future).

Reply:
Apart from the changes of this PR, it would be nice if Metals could load a Scala CLI project the same way it loads sbt or Mill ones (by running commands, that allow it to load the project via Bloop or BSP later on). *This* would involve calling `scala-cli setup-ide`. But that should rather be asked when Metals detects the build tool I think, not when it detects a script.

Reply:
That said, `TargetData` introduced in https://github.com/scalameta/metals/pull/3550 somehow paves the way towards proper multiple BSP servers support. Maybe it's time to try to add support for that 🤔 (it should be doable to make Ammonite write a `.bsp/ammonite.json` file too, that way both Ammonite and Scala CLI ad hoc BSP server stuff could be scrapped). But that's a bit beyond the scope of this PR.